src/storage/txn/commands/prewrite.rs

```diff
@@ -433,6 +433,30 @@ impl<K: PrewriteKind> Prewriter<K> {
         let mut final_min_commit_ts = TimeStamp::zero();
         let mut locks = Vec::new();
 
+        // Furthur check whether the prewrited transaction has been committed
+        // when encountering a WriteConflict or PessimisticLockNotFound error.
+        // This extra check manages to make prewrite idempotent after the transaction
+        // was committed.
+        // Note that this check cannot fully guarantee idempotence because an MVCC
+        // GC can remove the old committed records, then we cannot determine
+        // whether the transaction has been committed, so the error is still returned.
+        fn check_committed_record(
+            prewrite_result: MvccResult<(TimeStamp, OldValue)>,
+            reader: &mut SnapshotReader<impl Snapshot>,
+            key: &Key,
+        ) -> Result<(Vec<std::result::Result<(), StorageError>>, TimeStamp)> {
+            if let TxnCommitRecord::SingleRecord { commit_ts, .. } =
+                reader.get_txn_commit_record(key)?
+            {
+                info!("prewrited transaction has been committed";
+                    "start_ts" => reader.start_ts, "commit_ts" => commit_ts);
+                Ok((vec![], commit_ts))
+            } else {
+                prewrite_result?;
+                unreachable!()
```

Comment:
Why?

Reply:
I change how to organize the code. Hope this is more understandable.